privilege/privileges/cache.go

```diff
@@ -0,0 +1,273 @@
+
+type dbTableRow struct {
+	Host       string
+	DB         string
+	User       string
+	Privileges mysql.PrivilegeType
+}
+
+const dbTablePrivilegeMask = mysql.SelectPriv | mysql.InsertPriv | mysql.UpdatePriv | mysql.DeletePriv | mysql.CreatePriv | mysql.DropPriv | mysql.GrantPriv | mysql.IndexPriv | mysql.AlterPriv
+
+type tablesPrivTableRow struct {
+	Host       string
+	DB         string
+	User       string
+	TableName  string
+	Grantor    string
+	Timestamp  time.Time
+	TablePriv  mysql.PrivilegeType
+	ColumnPriv mysql.PrivilegeType
+}
+
+const tablePrivMask = mysql.SelectPriv | mysql.InsertPriv | mysql.UpdatePriv | mysql.DeletePriv | mysql.CreatePriv | mysql.DropPriv | mysql.GrantPriv | mysql.IndexPriv | mysql.AlterPriv
+
+const columnPrivMask = mysql.SelectPriv | mysql.InsertPriv | mysql.UpdatePriv
+
+type columnsPrivTableRow struct {
+	Host       string
+	DB         string
+	User       string
+	TableName  string
+	ColumnName string
+	Timestamp  time.Time
+	ColumnPriv mysql.PrivilegeType
+}
+
+// MySQLPrivilege is the in-memory cache of mysql privilege tables.
+type MySQLPrivilege struct {
+	User        []userTableRow
+	DB          []dbTableRow
+	TablesPriv  []tablesPrivTableRow
+	ColumnsPriv []columnsPrivTableRow
+}
+
+// LoadAll loads the tables from database to memory.
+func (p *MySQLPrivilege) LoadAll(ctx context.Context) error {
+	err := p.LoadUserTable(ctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	err = p.LoadDBTable(ctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	err = p.LoadTablesPrivTable(ctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	err = p.LoadColumnsPrivTable(ctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+// LoadUserTable loads the mysql.user table from database.
+func (p *MySQLPrivilege) LoadUserTable(ctx context.Context) error {
+	return p.loadTable(ctx, "select * from mysql.user order by host, user;", p.decodeUserTableRow)
+}
+
+// LoadDBTable loads the mysql.db table from database.
+func (p *MySQLPrivilege) LoadDBTable(ctx context.Context) error {
+	return p.loadTable(ctx, "select * from mysql.db order by host, db, user;", p.decodeDBTableRow)
+}
+
+// LoadTablesPrivTable loads the mysql.tables_priv table from database.
+func (p *MySQLPrivilege) LoadTablesPrivTable(ctx context.Context) error {
+	return p.loadTable(ctx, "select * from mysql.tables_priv", p.decodeTablesPrivTableRow)
+}
+
+// LoadColumnsPrivTable loads the mysql.columns_priv table from database.
+func (p *MySQLPrivilege) LoadColumnsPrivTable(ctx context.Context) error {
+	return p.loadTable(ctx, "select * from mysql.columns_priv", p.decodeColumnsPrivTableRow)
+}
+
+func (p *MySQLPrivilege) loadTable(ctx context.Context, sql string,
+	decodeTableRow func(*ast.Row, []*ast.ResultField) error) error {
```

Comment:
By the name, `tableRowDecoder` means an object, while `decodeTableRow`  means an action
An action is better here, see line https://github.com/pingcap/tidb/pull/2388/files#diff-bf44fbe39866829e4b66d20ae4c5f501R141